### PR TITLE
Implemented chaindata fetcher API

### PR DESCRIPTION
**Pull request by jeongkyun-oh**
## Proposed changes

- Implemented chaindata fetcher API methods

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: datasync/chaindatafetcher/chaindata_fetcher.go [hunk 1]
```diff
@@ -107,9 +127,102 @@ func (f *ChainDataFetcher) Stop() error {
...
+
+	f.started = true
 	return nil
 }
 
+func (f *ChainDataFetcher) stop() error {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	if !f.started {
+		return errors.New("the chaindata fetcher is already stopped")
```
Comment (ehnuje) [inline datasync/chaindatafetcher/chaindata_fetcher.go:152]: Below would be more appropriate.
```suggestion
		return errors.New("the chaindata fetcher is not running")
```

File: datasync/chaindatafetcher/chaindata_fetcher.go [hunk 2]
```diff
@@ -107,9 +127,102 @@ func (f *ChainDataFetcher) Stop() error {
 	logger.Info("wait for all goroutines to be terminated...")
 	f.wg.Wait()
 	logger.Info("terminated all goroutines for chaindatafetcher")
+	f.started = false
+	return nil
+}
+
+func (f *ChainDataFetcher) start() error {
```
Comment (ehnuje) [inline datasync/chaindatafetcher/chaindata_fetcher.go:134]: Do we need feature?

File: datasync/chaindatafetcher/chaindata_fetcher.go [hunk 3]
```diff
@@ -107,9 +127,102 @@ func (f *ChainDataFetcher) Stop() error {
...
+	}
+
+	for _, e := range events {
+		f.reqCh <- &request{
+			reqType: requestTypeTransaction,
+			event:   e,
+		}
+		// TODO-ChainDataFetcher add stop logic while processing the events.
+	}
+	return nil
```
Comment (ehnuje) [inline datasync/chaindatafetcher/chaindata_fetcher.go:181]: This won't be returned unless the whole request is finished.

File: datasync/chaindatafetcher/api.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+func (api *PublicChainDataFetcherAPI) Start() error {
+	return api.f.start()
+}
+
+func (api *PublicChainDataFetcherAPI) Stop() error {
+	return api.f.stop()
+}
+
+func (api *PublicChainDataFetcherAPI) StartWithRange(start, end int64, reqType uint) error {
```
Comment (KimKyungup) [inline datasync/chaindatafetcher/api.go:35]:
```suggestion
func (api *PublicChainDataFetcherAPI) StartWithRange(start, end uint64, reqType uint) error {
```

File: datasync/chaindatafetcher/chaindata_fetcher.go (hunk 3 above)
Comment (KimKyungup) [reply datasync/chaindatafetcher/chaindata_fetcher.go:181]: We can add a scheduler for Range job later.

File: datasync/chaindatafetcher/chaindata_fetcher.go [hunk 5]
```diff
@@ -107,9 +127,102 @@ func (f *ChainDataFetcher) Stop() error {
...
+}
+
+func (f *ChainDataFetcher) stopRange() error {
+	// TODO-ChainDataFetcher add logic for stopping
+	return nil
+}
+
+func (f *ChainDataFetcher) makeChainEvents(start, end int64) ([]blockchain.ChainEvent, error) {
```
Comment (KimKyungup) [inline datasync/chaindatafetcher/chaindata_fetcher.go:189]: In next time, please change the code that each handler makes the event itself.

File: datasync/chaindatafetcher/chaindata_fetcher.go [hunk 6]
```diff
@@ -92,24 +109,115 @@ func (f *ChainDataFetcher) Start(server p2p.Server) error {
...
 	close(f.stopCh)
 
 	logger.Info("wait for all goroutines to be terminated...")
 	f.wg.Wait()
 	logger.Info("terminated all goroutines for chaindatafetcher")
+	f.started = false
+	return nil
+}
+
+func (f *ChainDataFetcher) start() error {
```
Comment (aidan-kwon) [inline datasync/chaindatafetcher/chaindata_fetcher.go:137]: `start` and `Start` methods confuse me. How about changing `start` to `startFetching`?
Also, I think `start` can be called inside of `Start`.

Comment (ehnuje) [reply datasync/chaindatafetcher/chaindata_fetcher.go:137]: How about start, stop -> startFetching, stopFetching / startRange, stopRange -> startRangeFetching, stopRangeFetching?

Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/chaindata_fetcher.go]: Applied. df422607e7afac34e8b07c8daf66014632022461

File: datasync/chaindatafetcher/chaindata_fetcher.go (hunk 5 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/chaindata_fetcher.go]: I added comment. I will do this later on. 0bad014154444e6d3381154d41c8c131a1b9a085

File: datasync/chaindatafetcher/api.go (hunk 4 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/api.go]: Applied. 57f787b039a2142433a12be2701902c2791fa8e5

File: datasync/chaindatafetcher/chaindata_fetcher.go (hunk 3 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/chaindata_fetcher.go:188]: For now, I divided the making chain event one by one. 9442e828f8e13bcb92540579f2cdc432308a8aa7

File: datasync/chaindatafetcher/chaindata_fetcher.go (hunk 1 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/chaindata_fetcher.go]: Modified on 225a181fc813970d4525c5ced9664518b7889f07.

**Comment by jeongkyun-oh** [comment]
@KimKyungup @aidan-kwon @ehnuje Thanks for your reviews. Still many things to do, so I added TODO comments. I will work after merging.

File: datasync/chaindatafetcher/api.go [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+func (api *PublicChainDataFetcherAPI) Start() error {
+	return api.f.startFetching()
+}
+
+func (api *PublicChainDataFetcherAPI) Stop() error {
+	return api.f.stopFetching()
+}
+
+func (api *PublicChainDataFetcherAPI) StartWithRange(start, end uint64, reqType uint) error {
```
Comment (ehnuje) [inline datasync/chaindatafetcher/api.go:35]:
```suggestion
func (api *PublicChainDataFetcherAPI) StartRangeFetcing(start, end uint64, reqType uint) error {
```

Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/api.go]: Applied. 22f67aa

**Pull request merged**